proposals/NNNN-fix-expressible-by-string-interpolation.md

```diff
@@ -0,0 +1,785 @@
+trusting `init(stringInterpolation:)` to correct them later.
+
+It also relies on an undocumented parser quirk that could easily be 
+changed in the future. There are no tests in Swift 3 to ensure this 
+behavior does not change, and doing so might speed up string literal 
+construction.
+
+### Defect 3: Cannot constrain interpolated types
+
+Types like `SQLKit.SQLStatement` only support certain types 
+being interpolated—`String`, `Data`, `Date`, `Bool`, `BinaryInteger`s, 
+`FloatingPoint`s, and `Decimal`s, but not `UIView`s, `Array`s, 
+or `User`s. Even `LocalizableString`, if implemented using 
+`String(format:)` for Objective-C compatibility, can only handle 
+`CVarArg` interpolations. `ExpressibleByStringInterpolation` cannot 
+easily express these kind of constraints.
+
+Like defect 2, there *is* a hack that sort of works around this. Swift 
+does not only use the specific `init<T>(stringInterpolationSegment: T)` 
+implementation listed in the protocol; it will actually consider all of 
+its overloads as well. So you can implement the required initializer, 
+but mark it `@available(*, unavailable: 0.0)`, and then overload it 
+without the annotation for the types you want to support. The compiler 
+will emit an error if you try to use the unconstrained generic 
+overload. Exploiting this, however, is a pretty dirty trick.
+
+### Defect 4: No formatting syntax
+
+When interpolating a value into a string, it's often necessary to 
+control details of how it's formatted—think of the radix of an 
+integer or the precision of a floating-point value. Swift's 
+standard way to do this is to write an expression which generates 
+a formatted string, but this adds extra verbiage to the 
+interpolated expression and (partially due to defect 3) limits 
+the ability of types to override formatting.
+
+When considering this problem, keep in mind that "formatting" is a very 
+broad category of actions. For instance, `SQLStatement` might want to 
+support inserting raw SQL from a `String`; that's a form of formatting. 
+If `SanitizedHTML` wanted to wrap interpolated `Date`s in HTML5 
+`<time>` tags, that would be a form of formatting too. 
+
+## Proposed solution
+
+### Turn interpolation syntax into an initializer call
+
+We address defects 3 and 4 by changing the interpretation of the 
+`\(expr)` syntax from creating an expression to creating an initializer 
+call. Specifically, it will call an initializer on a new 
+`ExpressibleByStringInterpolation` associated type called 
+`StringInterpolationType`. The contents of the parentheses are treated 
+as a parameter list.
+
+When `StringInterpolationType` is the built-in `String` type, this will 
+leverage existing initializers on `String`:
+
+```string
+print("\(number) in hex is \(number, radix: 16)")
+print("Debug info: \(reflecting: object)")
+print("Disclaimer: \(contentsOfURL: Bundle.main.url(forResource: "disclaimer", withExtension: "txt")!)")
```

Comment:
This expression is similar to `String.init(contentsOf: URL) throws` which requires error handling.

<https://github.com/apple/swift/blob/9b7cd360cec9cd4644ce7dbfe76639af8537f1b8/stdlib/public/SDK/Foundation/NSStringAPI.swift#L879-L884>